provider/adguardhome/adguardhome.go

```diff
@@ -0,0 +1,194 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package adguardhome
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"os"
+	"regexp"
+
+	adguard "github.com/markussiebert/go-adguardhome-client/client"
+	log "github.com/sirupsen/logrus"
+
+	"sigs.k8s.io/external-dns/endpoint"
+	"sigs.k8s.io/external-dns/plan"
+	"sigs.k8s.io/external-dns/provider"
+)
+
+// The regex hat should identify and parse all custom rules in adguard home that are used to handle dns records
+var regex = regexp.MustCompile(`(?P<dnsName>[a-zA-Z\*.-]+)\$dnstype=(?P<recordType>[A-Z]+),dnsrewrite=NOERROR;[A-Z]+;(?P<target>.*)`)
+
+// AdguardHomeProvider is an implementation of Provider for Vultr DNS.
+type AdguardHomeProvider struct {
+	provider.BaseProvider
+	client adguard.ClientWithResponses
+
+	domainFilter endpoint.DomainFilter
+	DryRun       bool
+}
+
+// Prepare basicAuth with credentials
+func basicAuth(username, password string) string {
+	auth := username + ":" + password
+	return "Basic " + base64.StdEncoding.EncodeToString([]byte(auth))
+}
+
+// NewAdguardHomeProvider initializes a new Vultr BNS based provider
+func NewAdguardHomeProvider(domainFilter endpoint.DomainFilter, dryRun bool) (*AdguardHomeProvider, error) {
+	adguardHomeURL, adguardHomeUrlOk := os.LookupEnv("ADGUARD_HOME_URL")
+	if !adguardHomeUrlOk {
```

Comment:
Might want to document that the AGH URL passed here needs to be suffixed with /control/ for a regular AGH deployment. Since that's where the APIs are under.

Reply:
I got confused trying to troubleshoot that with a very unhelpful error about a nil reference.

Reply:
You are right - it took me some time to figure Out the right path...